### PR TITLE
[bitnami/postgresql] Fix conf.d permissions with postgresqlextendedConf

### PR DESCRIPTION
**Pull request by strainovic**
`postgresqlextendedConf` creates a directory under conf/conf.d

When we install this chart for the first time, everything works fine until we try to upgrade.

```bash
+ mkdir -p /bitnami/postgresql/conf /bitnami/postgresql/data
+ chmod 700 /bitnami/postgresql/conf /bitnami/postgresql/data
+ find /bitnami/postgresql -mindepth 1 -maxdepth 1 -not -name .snapshot -not -name lost+found
+ id -u
+ id -G
+ cut -d   -f2
+ xargs chown -R 1000540000:1000540000
chown: changing ownership of '/bitnami/postgresql/conf/conf.d': Operation not permitted
```

image

My values.yaml override:

```yaml
  volumePermissions:
    enabled: true
    securityContext:
      runAsUser: "auto"
  securityContext:
    enabled: false
  shmVolume:
    chmod:
      enabled: false
```

I tried to add `postgresqlConfiguration` which creates a file under conf directory and it works fine.

Related to: https://github.com/bitnami/charts/pull/2141

### CODE REVIEW

**Comment by strainovic** [comment]
Hm, it happened again... strange.

```
+ mkdir -p /bitnami/postgresql/conf/conf.d /bitnami/postgresql/data
+ chmod 700 /bitnami/postgresql/conf/conf.d /bitnami/postgresql/data
chmod: changing permissions of '/bitnami/postgresql/conf/conf.d': Operation not permitted
+ find /bitnami/postgresql -mindepth 1 -maxdepth 1 -not -name .snapshot -not -name lost+found
+ id -u
+ id -G
+ cut -d   -f2
+ xargs chown -R 1000540000:1000540000
chown: changing ownership of '/bitnami/postgresql/conf/conf.d': Operation not permitted
chown: changing ownership of '/bitnami/postgresql/conf': Operation not permitted
```

**Comment by juan131** [comment]
Hi @strainovic 

When using the `postgresqlextendedConf` parameter, we're mounting a read-only volume with the configmap's content in the container. This "read-only" volume doesn't accept any kind of write operations, nor operations attempting to change the ownership/permissions.

Therefore, what we need to do is excluding this directory from the "find" command:

```diff
- find {{ .Values.persistence.mountPath }} -mindepth 1 -maxdepth 1 -not -name ".snapshot" -not -name "lost+found"
+ find {{ .Values.persistence.mountPath }} -mindepth 1 -maxdepth 1 -not -name "conf.d" -not -name ".snapshot" -not -name "lost+found"
```

Could you update the PR addressing this suggestion and test it?

**Comment by strainovic** [comment]
Hi @juan131,

I would exclude `conf` folder as it is on the first level of depth. There we mount `postgresql.conf` and inside it, we can have `conf.d` with other `*.conf` files.

Btw. I cannot figure out why do we need that init container? I tested with and without it and I see no difference in OpenShift. In both cases, it works fine and permissions are correct.

I'm confused because I cannot find the case when we must enable it.

I use vSphere storage provider pod PVCs.

File: bitnami/postgresql/templates/statefulset-slaves.yaml [hunk 1]
```diff
@@ -83,9 +83,9 @@ spec:
             - -cx
             - |
               {{ if .Values.persistence.enabled }}
-              mkdir -p {{ .Values.persistence.mountPath }}/conf {{ .Values.persistence.mountPath }}/data
-              chmod 700 {{ .Values.persistence.mountPath }}/conf {{ .Values.persistence.mountPath }}/data
```
Comment (juan131) [inline bitnami/postgresql/templates/statefulset-slaves.yaml:87]: When the configuration file is not mounted, it's autogenerated and writing permissions will be required. Suggestion:

- Create this macro in the `_helpers.tpl`:

```
{{/*
Return true if a configmap should be mounted with PostgreSQL configuration
*/}}
{{- define "postgresql.mountConfigmap" -}}
{{- if or (.Files.Glob "files/postgresql.conf") (.Files.Glob "files/pg_hba.conf") .Values.postgresqlConfiguration .Values.pgHbaConfiguration .Values.configurationConfigMap }}
    {- true -}}
{{- end -}}
{{- end -}}
```

- Use it here:

```
mkdir -p {{ .Values.persistence.mountPath }}/data {{ if not (include "postgresql.mountConfigmap" .) }} {{ .Values.persistence.mountPath }}/conf {{ end }}
chmod 700 {{ .Values.persistence.mountPath }}/data {{ if not (include "postgresql.mountConfigmap" .) }} {{ .Values.persistence.mountPath }}/conf {{ end }}
find {{ .Values.persistence.mountPath }} -mindepth 1 -maxdepth 1 {{ if not (include "postgresql.mountConfigmap" .) }} -not -name "conf" {{ end }} -not -name ".snapshot" -not -name "lost+found" | \
```

Comment (strainovic) [reply bitnami/postgresql/templates/statefulset-slaves.yaml:87]: I added it but had to change conditions for mkdir and chmod commands.

```yaml
  postgresqlConfiguration:
    listenAddresses: "'*'"
    port: "'5432'"
  postgresqlExtendedConf:
    maxConnections: 1024
  volumePermissions:
    enabled: true
    securityContext:
      runAsUser: "auto"
  securityContext:
    enabled: false
  shmVolume:
    chmod:
      enabled: false
```

I did some tests and it works fine. I think we are good to go :)

**Comment by juan131** [comment]
Waiting for CI/CD to test your PR and update image tags.

**Comment by strainovic** [comment]
Hi @juan131,

Thank you for accepting my PR.

Did you notice a bug in your CI/CD tool?

I've bumped up Chart.yaml to version 8.9.2 when I started working on this PR. In the meanwhile, someone else did the same and its PR was merged back to the master before you merged my PR.

Now, when I pull bitnami/postgresql chart version 8.9.2 it does not contain my changes, but only changes from the previous PR. Isn't it weird?

statefulset.yaml from helm chart repository version 8.9.2

image

statefulset.yaml from the master branch with the same version 8.9.2

image

Btw, I think you made a mistake with the postgres-exporter tag, it should be r91, not r9.

Thanks!

**Comment by carrodher** [comment]
Hi, yes, during the review process of this PR another PR was merged bumping the chart version from `8.9.1` to `8.9.2`. The chart version should have been bumped from `8.9.2` to `8.9.3` in this PR, we as reviewers, the CI/CD system nor GitHub during the merge detected the issue so this PR overwrite the existing changes. But this "issue" should affect the previous PR and not this one.

You should see your changes in the existing `8.9.2` version, in fact, this PR has overwritten the existing changes. I think what happens is that you have not updated the Helm repository and you are seeing the previous changes (those ones overwritten by this PR). Updating the repository your changes are present in the `8.9.2` version:

```
▶ helm repo update
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "bitnami" chart repository
Update Complete. ⎈ Happy Helming!⎈

▶ helm fetch bitnami/postgresql --version 8.9.2

▶ tar xzf postgresql-8.9.2.tgz && cd postgresql

▶ ag 'mkdir -p {{ .Values.persistence.mountPath }}/data'
templates/statefulset-slaves.yaml
86:              mkdir -p {{ .Values.persistence.mountPath }}/data {{- if (include "postgresql.mountConfigurationCM" .) }} {{ .Values.persistence.mountPath }}/conf {{- end }}

templates/statefulset.yaml
86:              mkdir -p {{ .Values.persistence.mountPath }}/data {{- if (include "postgresql.mountConfigurationCM" .) }} {{ .Values.persistence.mountPath }}/conf {{- end }}
```

The same for the exporter tag, the `values.yaml` in that chart is using the `-r91`
```
▶ ag 'tag: ' values.yaml
18:  tag: 11.7.0-debian-10-r86
53:    tag: buster
520:    tag: 0.8.0-debian-10-r91
```

Can you confirm if your changes are present from the `8.9.2` version once updated the repository?

**Comment by strainovic** [comment]
Hi @carrodher,

Actually, I did helm repo up before I fetched version 8.9.2. I did it again and it pulled the correct newer version.

Maybe CI should report version override to prevent this to happen again?

Thanks

**Comment by carrodher** [comment]
Glad to see that it's working now for you.

We are working to improve the CI/CD pipeline, focused on user's contributions via GH PRs, so definitively yes, we are consideration it. Thanks for the suggestion.

**Pull request merged**